File: tests/pattern.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { parsePattern } from "../src/pattern.js";

test("parses wildcard", () => {
  assert.deepEqual(parsePattern("*"), { raw: "*", isWildcard: true, segments: [] });
});

test("parses static root", () => {
  const result = parsePattern("/");
  assert.equal(result.isWildcard, false);
  assert.deepEqual(result.segments, []);
});

test("parses static segments", () => {
  const result = parsePattern("/articles");
  assert.deepEqual(result.segments, [{ type: "static", value: "articles" }]);
});

test("parses dynamic segments", () => {
  const result = parsePattern("/articles/:slug");
  assert.deepEqual(result.segments, [
    { type: "static", value: "articles" },
    { type: "param", name: "slug" },
  ]);
});

test("parses multiple dynamic segments", () => {
  const result = parsePattern("/users/:id/posts/:postId");
  assert.deepEqual(result.segments, [
    { type: "static", value: "users" },
    { type: "param", name: "id" },
    { type: "static", value: "posts" },
    { type: "param", name: "postId" },
  ]);
});

File: go/pattern_test.go
package roution

import (
	"reflect"
	"testing"
)

func TestParsePattern(t *testing.T) {
	tests := []struct {
		name    string
		pattern string
		want    ParsedPattern
	}{
		{
			name:    "parses wildcard",
			pattern: "*",
			want: ParsedPattern{
				Raw:        "*",
				IsWildCard: true,
				Segments:   []Segment{},
			},
		},
		{
			name:    "parses static root",
			pattern: "/",
			want: ParsedPattern{
				Raw:      "/",
				Segments: []Segment{},
			},
		},
		{
			name:    "parses static segments",
			pattern: "/articles",
			want: ParsedPattern{
				Raw: "/articles",
				Segments: []Segment{
					{
						Type:  SegmentStatic,
						Value: "articles",
					},
				},
			},
		},
		{
			name:    "parses dynamic segments",
			pattern: "/articles/:slug",
			want: ParsedPattern{
				Raw: "/articles/:slug",
				Segments: []Segment{
					{
						Type:  SegmentStatic,
						Value: "articles",
					},
					{
						Type: SegmentParam,
						Name: "slug",
					},
				},
			},
		},
		{
			name:    "parses multiple dynamic segments",
			pattern: "/users/:id/posts/:postId",
			want: ParsedPattern{
				Raw: "/users/:id/posts/:postId",
				Segments: []Segment{
					{
						Type:  SegmentStatic,
						Value: "users",
					},
					{
						Type: SegmentParam,
						Name: "id",
					},
					{
						Type:  SegmentStatic,
						Value: "posts",
					},
					{
						Type: SegmentParam,
						Name: "postId",
					},
				},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ParsePattern(tt.pattern)

			if !reflect.DeepEqual(got, tt.want) {
				t.Fatalf(
					"ParsePattern(%q)\n\n got: %#v\nwant: %#v",
					tt.pattern,
					got,
					tt.want,
				)
			}
		})
	}
}